phoenix-core/src/test/java/org/apache/phoenix/jdbc/FailoverPhoenixConnectionTest.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.phoenix.exception.FailoverSQLException;
+import org.apache.phoenix.exception.SQLExceptionCode;
+import org.apache.phoenix.jdbc.HighAvailabilityGroup.HAGroupInfo;
+
+/**
+ * Unit test for {@link FailoverPhoenixConnection}.
+ *
+ * @see FailoverPhoenixConnectionIT
+ */
+public class FailoverPhoenixConnectionTest {
+    private static final Logger LOG = LoggerFactory.getLogger(FailoverPhoenixConnectionTest.class);
+
+    @Mock PhoenixConnection connection1;
+    @Mock PhoenixConnection connection2;
+    @Mock HighAvailabilityGroup haGroup;
+
+    final HAGroupInfo haGroupInfo = new HAGroupInfo("fake", "zk1", "zk2");
+    FailoverPhoenixConnection failoverConnection; // this connection itself is not mocked or spied.
+
+    @Before
+    public void init() throws SQLException {
+        MockitoAnnotations.initMocks(this);
+        when(haGroup.getGroupInfo()).thenReturn(haGroupInfo);
+        when(haGroup.connectActive(any(Properties.class))).thenReturn(connection1);
+
+        failoverConnection = new FailoverPhoenixConnection(haGroup, new Properties());
+    }
+
+    /**
+     * Test helper method {@link FailoverPhoenixConnection#wrapActionDuringFailover}.
+     */
+    @Test
+    public void testWrapActionDuringFailover() throws SQLException {
+        // Test SupplierWithSQLException which returns a value
+        String str = "Hello, World!";
+        assertEquals(str, failoverConnection.wrapActionDuringFailover(() -> str));
+
+        // Test RunWithSQLException which does not return value
+        final AtomicInteger counter = new AtomicInteger(0);
+        failoverConnection.wrapActionDuringFailover(counter::incrementAndGet);
+        assertEquals(1, counter.get());
+    }
+
+    /**
+     * Test that after calling failover(), the old connection got closed with FailoverSQLException,
+     * and a new Phoenix connection is opened.
+     */
+    @Test
+    public void testFailover() throws SQLException {
+        // Make HAGroup return a different phoenix connection when it gets called next time
+        when(haGroup.connectActive(any(Properties.class))).thenReturn(connection2);
+
+        // explicit call failover
+        failoverConnection.failover(1000L);
+
+        // The old connection should have been closed due to failover
+        verify(connection1, times(1)).close(any(FailoverSQLException.class));
+        // A new Phoenix connection is wrapped underneath
+        assertEquals(connection2, failoverConnection.getWrappedConnection());
+    }
+
+    /**
+     * Test static {@link FailoverPhoenixConnection#failover(Connection, long)} method.
+     */
+    @Test
+    public void testFailoverStatic() throws SQLException {
+        try {
+            FailoverPhoenixConnection.failover(connection1, 1000L);
+            fail("Should have failed since plain phoenix connection can not failover!");
+        } catch (SQLException e) {
+            LOG.info("Got expected exception when trying to failover on non-HA connection", e);
```

Comment:
nit: can we match the SQLExceptionCode to make sure it's the right error?

Reply:
Done